lib/spack/spack/installer.py

```diff
@@ -1641,6 +1714,16 @@ def build_process(pkg, kwargs):
     skip_patch = kwargs.get('skip_patch', False)
     unmodified_env = kwargs.get('unmodified_env', {})
     verbose = kwargs.get('verbose', False)
+    use_monitor = kwargs.get('use_monitor', False)
+    monitor = None
+
+    # Instantiate a monitor for the worker, if requested
+    if use_monitor:
```

Comment:
similarly curious here whether monitors could be implemented via hooks -- can we have a general set of hooks here and not an `if monitor` for this feature (and the next feature and the next one ...)

Reply:
I'm not familiar with how hooks work, is there an example somewhere?

Reply:
## Did somebody say they wanted a hook?
![image](https://user-images.githubusercontent.com/814322/111557156-c1e45700-8751-11eb-8508-982c62caa6dd.png)

Reply:
Look here: https://github.com/spack/spack/blob/develop/lib/spack/spack/hooks/__init__.py and at all the modules in that directory.

Basically, instead of sticking monitor-specific calls here, I think `on_install_failure(exception)` or something like that should be a new hook for these spots in the code.  Then we can write other things that need to handle the same events.

A hook is just a callback.  It's yet another way to do the observer pattern.

Reply:
Also the idea is to get all the monitor code in one place -- right now it's scattered all over installer.

Reply:
Okay so two questions:

1. If the hook just accepts an exception, how do we also give it other variables like monitor preferences from the user?
2. Given something like the monitor, how do we instantiate it once (and then use over again) so we don't have to do it every time there is a hook (related to 1.)

Reply:
1. Not tied to the arguments -- that's up to you.  I think making it take a spec as well makes a lot of sense.
2. In the hook module, I think you should create an object and store it when installation starts (so the first hook would create the object).  Just spitballing here, but I might store in a `dict` in the hook module, keyed by spec.